### PR TITLE
preserve TMPDIR and HOSTALIASES across snap-confine invocation (LP: #1682308)

### PR DESCRIPTION
**Pull request by mwhudson**
Because snap-confine is setuid, the dynamic loader strips out certain
environment variables when starting the process. Work around this by renaming
them before invoking snap-confine and renaming them back in snap-exec.

### CODE REVIEW

**Comment by mwhudson** [comment]
I'm not sure but perhaps this is only something one would want for classic snaps?

File: snap/snapenv/snapenv.go [hunk 1]
```diff
@@ -106,8 +106,17 @@ func userEnv(info *snap.Info, home string) map[string]string {
 	return result
 }
 
-// envMap creates a map from the given environment string list, e.g. the
-// list returned from os.Environ()
+var unsafeEnv = map[string]bool{
+	"HOSTALIASES": true,
+	"TMPDIR":      true,
```
Comment (zyga) [inline snap/snapenv/snapenv.go]: I'm not sure TMPDIR should be preserved. The designated directory may not exist inside the execution environment. What is your motivation for preserving it?

File: snap/snapenv/snapenv.go [hunk 2]
```diff
@@ -106,8 +106,17 @@ func userEnv(info *snap.Info, home string) map[string]string {
 	return result
 }
 
-// envMap creates a map from the given environment string list, e.g. the
-// list returned from os.Environ()
+var unsafeEnv = map[string]bool{
+	"HOSTALIASES": true,
```
Comment (zyga) [inline snap/snapenv/snapenv.go]: If the user sets HOSTALIASES to a directory that is not readable due to the sandbox shoud this be still preserved?

**Comment by mwhudson** [comment]
The motivation is that this keeps confusing users of my Go snap: https://bugs.launchpad.net/snapd/+bug/1682308

I think your questions indicate the answer to my question: this should only be done for a classic snap, and it turns out that that's really easy (pushed).

Here is the full list of variables glibc strips out: https://sourceware.org/git/?p=glibc.git;a=blob_plain;f=sysdeps/generic/unsecvars.h;hb=HEAD -- I guess it would make sense to save some more of these, maybe even all of them?

**Comment by chipaca** [comment]
LGTM, but, should we ask Jamie to review this?

**Comment by mwhudson** [comment]
I fixed the conflicts by amending the second commit to not touch the changelog (it was a mistake to commit it in the first place)

File: snap/snapenv/snapenv.go [hunk 3]
```diff
@@ -106,16 +110,29 @@ func userEnv(info *snap.Info, home string) map[string]string {
 	return result
 }
 
-// envMap creates a map from the given environment string list, e.g. the
-// list returned from os.Environ()
-func envMap(env []string) map[string]string {
+var unsafeEnv = map[string]bool{
+	"HOSTALIASES": true,
+	"TMPDIR":      true,
```
Comment (niemeyer) [inline snap/snapenv/snapenv.go]: Sounds good to preserve all of them.

File: snap/snapenv/snapenv.go [hunk 4]
```diff
@@ -106,16 +110,29 @@ func userEnv(info *snap.Info, home string) map[string]string {
...
+const PreservedUnsafePrefix = "SNAP_SAVED_"
+
+// envMap creates a map from the given environment string list,
+// e.g. the list returned from os.Environ(). If preserveUnsafeVars
+// rename some variables that will be stripped out by the dynamic
+// linker executing the setuid snap-confine by prepending their names
+// with PreservedUnsafePrefix.
+func envMap(env []string, preserveUnsafeVars bool) map[string]string {
 	envMap := map[string]string{}
 	for _, kv := range env {
```
Comment (niemeyer) [inline snap/snapenv/snapenv.go]:
```
if strings.HasPrefix(kv, PreservedUnsafePrefix) {
        continue
}
```

File: snap/snapenv/snapenv.go (hunk 3 above)
Comment (mwhudson) [reply snap/snapenv/snapenv.go]: Done

File: snap/snapenv/snapenv.go (hunk 4 above)
Comment (mwhudson) [reply snap/snapenv/snapenv.go]: It took me a moment to realize why you wanted this so I added a comment too.

File: snap/snapenv/snapenv.go [hunk 5]
```diff
@@ -106,16 +110,60 @@ func userEnv(info *snap.Info, home string) map[string]string {
...
+}
+
+const PreservedUnsafePrefix = "SNAP_SAVED_"
+
+// envMap creates a map from the given environment string list,
+// e.g. the list returned from os.Environ(). If preserveUnsafeVars
+// rename variables that will be stripped out by the dynamic linker
+// executing the setuid snap-confine by prepending their names with
+// PreservedUnsafePrefix.
+func envMap(env []string, preserveUnsafeVars bool) map[string]string {
```
Comment (mvo5) [inline snap/snapenv/snapenv.go]: nitpick - I'm not a fan of this boolean here. It makes it hard to reason about what the function is doing without looking at the signature. I.e.: what does `envMap(env, true)` mean? I would prefer a flag argument and a name, maybe something like `envMap(env, {0, preserveUnsafeVars])` ?

Comment (mwhudson) [reply snap/snapenv/snapenv.go]: I'm not 100% what you mean, but see what I pushed?

Comment (mvo5) [reply snap/snapenv/snapenv.go]: Yeah, this is what I had in mind, thanks for this! Sorry that I was hard to parse.

**Comment by mvo5** [comment]
Code looks nice, thank you! I would love to see some sort of test for this, ideally both unit and spread. Please let us know if you would prefer help with this.

**Comment by mwhudson** [comment]
I've pushed some unit tests and a spread test. I haven't run the spread test locally so if it doesn't work I guess I'd like some help with that :)

File: tests/main/confinement-classic/task.yaml [hunk 6]
```diff
@@ -17,6 +17,8 @@ execute: |
     run_install --classic
     $SNAPMOUNTDIR/bin/test-snapd-hello-classic | MATCH 'Hello Classic!'
 
+    TMPDIR=/tmpdir $SNAPMOUNTDIR/bin/test-snapd-hello-classic t | MATCH TMPDIR=/tmpdir
```
Comment (mwhudson) [inline tests/main/confinement-classic/task.yaml]: This fails with "/bin/bash: line 65: SNAPMOUNTDIR: unbound variable" which doesn't make any sense to me. Also the instructions in HACKING.md for running the spread tests don't work. Help?

Comment (zyga) [reply tests/main/confinement-classic/task.yaml]: If you have the a key for linode API access (aka spread key) you can run this test with:

`spread -debug -v linode:ubuntu-16.04-64:tests/main/confinement-classic`

(substitute target environment if you want other OS)

I would also suggest to cut on the number of workers to just one (see spread.yaml) as otherwise debugging is weird and you waste machines that do nothing useful.

File: tests/main/confinement-classic/task.yaml [hunk 7]
```diff
@@ -21,7 +21,7 @@ execute: |
     run_install --classic
     $SNAP_MOUNT_DIR/bin/test-snapd-hello-classic | MATCH 'Hello Classic!'
 
-    TMPDIR=/tmpdir $SNAPMOUNTDIR/bin/test-snapd-hello-classic t | MATCH TMPDIR=/tmpdir
+    TMPDIR=/tmpdir $SNAP_MOUNT_DIR/bin/test-snapd-hello-classic t | MATCH TMPDIR=/tmpdir
```
Comment (zyga) [inline tests/main/confinement-classic/task.yaml]: Thank you

**Comment by mvo5** [comment]
This fails in the unit tests: 
```
coverage: 50.0% of statements
ok  	github.com/snapcore/snapd/snap/snapdir	0.004s
# github.com/snapcore/snapd/snap/snapenv
snap/snapenv/snapenv_test.go:195:9: undefined: val
snap/snapenv/snapenv_test.go:197:11: undefined: val
```
failure looks real.

File: snap/snapenv/snapenv.go [hunk 8]
```diff
@@ -31,18 +31,33 @@ import (
 	"github.com/snapcore/snapd/snap"
 )
 
+type preserveUnsafeEnvFlag int8
+
+const (
+	noPreserve preserveUnsafeEnvFlag = iota
```
Comment (zyga) [inline snap/snapenv/snapenv.go]: `noPreserve` and `doPreserve` differ by one character. With my ageing eyes I find it hard to distinguish easily. How about simply `preserve` and `notPreserve` (other ideas welcome)

File: snap/snapenv/snapenv.go [hunk 9]
```diff
@@ -113,16 +128,60 @@ func userEnv(info *snap.Info, home string) map[string]string {
 	return result
 }
 
-// envMap creates a map from the given environment string list, e.g. the
-// list returned from os.Environ()
-func envMap(env []string) map[string]string {
+// Environment variables glibc strips out when running a setuid binary.
+// Taken from https://sourceware.org/git/?p=glibc.git;a=blob_plain;f=sysdeps/generic/unsecvars.h;hb=HEAD
+var unsafeEnv = map[string]bool{
```
Comment (zyga) [inline snap/snapenv/snapenv.go]: This would be a nice thing to go generate eventually. No need to do it now but maybe add a comment.

**Comment by mwhudson** [comment]
> This looks ok to me overall, but please address my comment on renaming the unit tests.

Ack.

> Also, while we test in spread that TMPDIR makes it through to the classic snap, we should also test that TMPDIR doesn't make it through for non-classic.

OK, but I think I would also argue that if non-preservation of TMPDIR etc is important, we should probably filter them out ourselves (which would be easy to do in this branch, I guess) rather than relying on the dynamic loader doing it, as e.g. is doesn't look like musl does the same stripping or if for some reason we shipped a statically linked snap-confine there wouldn't be a dynamic loader at all.

Do you think I should change envMap to filter out "unsafe" variables in the non-classic case?

> In terms of the approach, I think it is ok because the actual setuid binary (snap-confine) will still have these variables stripped (ie, we aren't adding to the attack surface for snap-confine). If the classic snap shipped a setuid binary, it would still be stripped at exec().
> 
> Furthermore, a classic snap is intended to work more with the system settings so preserving (at least some of these) variables for classic makes sense. That said, it seems to me since we are preserving the various LD_* variables, this PR may add confusion for users and/or may complicate matters related to https://forum.snapcraft.io/t/classic-snaps-failing-on-ubuntu-17-10/2324/7. I've not studied that issue closely so will not block on this point, but please consider it before merging.

Although the only variable my users actually seem to care about is TMPDIR, I think I would argue that not allowing these variables through is more confusing than blocking them. My sympathy for people who set LD_LIBRARY_PATH without knowing what they are doing is a bit limited.

> For completeness and posterity, it is also worth noting that AppArmor looks at AT_SECURE with the profile transitions for Px, Cx and Ux. This PR will not affect that, since any exec transitions will still have the variables stripped at the time of the exec(). Specifically for snappy, classic snap policy intentionally has 'pix' for profile transitions (ie, to not strip these variables when the snap executes other binaries), and a non-snap application that has an apparmor profile that is executing a classic snap that has a Px/Cx/Ux rule for it will correctly still have these vars cleared on profile transition (just as the local admin specified).

**Comment by mwhudson** [comment]
> > That said, it seems to me since we are preserving the various LD_* variables, this PR may add confusion for users and/or may complicate matters related to https://forum.snapcraft.io/t/classic-snaps-failing-on-ubuntu-17-10/2324/7. I've not studied that issue closely so will not block on this point, but please consider it before merging.

> Although the only variable my users actually seem to care about is TMPDIR, I think I would argue that not allowing these variables through is more confusing than blocking them. My sympathy for people who set LD_LIBRARY_PATH without knowing what they are doing is a bit limited.

In addition, now that I've read that thread, the solution being proposed there in effect means that LD_LIBRARY_PATH will be ignored for binaries in classic snaps anyway.
